lib/cloud/gcp/imds.go

```diff
@@ -43,6 +45,80 @@ func (rt contextRoundTripper) RoundTrip(req *http.Request) (*http.Response, erro
 	return resp, trace.Wrap(err)
 }
 
+type metadataGetter func(ctx context.Context, path string) (string, error)
+
+// InstanceMetadataClient is a client for GCP instance metadata.
+type InstanceMetadataClient struct {
+	instancesClient InstancesClient
+	getMetadata     metadataGetter
+}
+
+// NewInstanceMetadataClient creates a new instance metadata client.
+func NewInstanceMetadataClient(ctx context.Context) (*InstanceMetadataClient, error) {
+	instancesClient, err := NewInstancesClient(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &InstanceMetadataClient{
+		instancesClient: instancesClient,
+		getMetadata:     getMetadata,
+	}, nil
+}
+
+// IsAvailable checks if instance metadata is available.
+func (client *InstanceMetadataClient) IsAvailable(ctx context.Context) bool {
+	instanceData, err := client.getMetadata(ctx, "instance")
+	return err == nil && instanceData != ""
+}
+
+// GetTags gets all of the GCP instance's labels (note: these are separate from
+// its tags, which we do not use).
+func (client *InstanceMetadataClient) GetTags(ctx context.Context) (map[string]string, error) {
+	projectID, err := client.getMetadata(ctx, "project/project-id")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	fullZone, err := client.getMetadata(ctx, "instance/zone")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// zone is formatted as projects/PROJECT_NUM/zones/ZONE, we just need the last part
+	zoneParts := strings.Split(fullZone, "/")
+	zone := zoneParts[len(zoneParts)-1]
+	name, err := client.getMetadata(ctx, "instance/name")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	inst, err := client.instancesClient.GetInstance(ctx, &InstanceRequest{
```

Comment:
are machines automatically given this permission to get themselves or does it require an extra permission?

Reply:
It requires extra permissions, which I'll add to the docs in a separate PR.